File: XMLCheck/Handler/BaseXmlParser.cs
using System.Globalization;
using System.Xml.Linq;

namespace XMLCheck.Handler
{
    public abstract class BaseXmlParser
    {
        protected XDocument? Document;
        protected XElement? Root;

        protected bool Load(string xmlContent)
        {
            if (string.IsNullOrWhiteSpace(xmlContent))
                return false;

            try
            {
                // Xử lý BOM + parse
                xmlContent = xmlContent.Trim('\uFEFF', '\u200B', '\u0000');

                Document = XDocument.Parse(xmlContent);
                Root = Document.Root;

                return Root != null;
            }
            catch
            {
                return false;
            }
        }

        // =========================
        // HELPER CORE (CHUẨN HOÁ)
        // =========================

        protected string? S(string key)
        {
            return Root?
                .Descendants()
                .FirstOrDefault(e => e.Name.LocalName == key)
                ?.Value
                ?.Trim();
        }

        protected int? I(string key)
        {
            return int.TryParse(S(key), out var v) ? v : null;
        }

        protected decimal? D(string key)
        {
            return decimal.TryParse(
                S(key),
                NumberStyles.Any,
                CultureInfo.InvariantCulture,
                out var v
            ) ? v : null;
        }

        protected DateTime? DT(string key)
        {
            var raw = S(key);
            if (string.IsNullOrWhiteSpace(raw)) return null;

            string[] formats =
            {
                "yyyyMMdd",
                "yyyyMMddHHmm",
                "yyyyMMddHHmmss"
            };

            return DateTime.TryParseExact(
                raw,
                formats,
                CultureInfo.InvariantCulture,
                DateTimeStyles.None,
                out var d
            ) ? d : null;
        }

        // =========================
        // HELPER ROOT CHECK
        // =========================

        protected bool IsRoot(string expectedLocalName)
        {
            if (Root == null) return false;

            var name = Root.Name.LocalName
                .Trim('\uFEFF')
                .Trim();

            return name == expectedLocalName;
        }
    }
}

File: XMLValidator_v2/src/parsers/base-xml.parser.ts
import { DOMParser } from '@xmldom/xmldom';
import { parseExactDate } from '../utils/date.helper';

export abstract class BaseXmlParser {
  protected document: any = null;
  protected root: any = null;

  public abstract get loaiHoSo(): string;
  public abstract parse(xmlContent: string): any;

  protected load(xmlContent: string): boolean {
    if (!xmlContent) return false;

    try {
      // Handle BOM characters
      const cleanedXml = xmlContent.replace(/^[\uFEFF\u200B\u0000]+/, '').trim();
      
      const parser = new DOMParser({
        errorHandler: {
          warning: () => {},
          error: () => {},
          fatalError: () => {},
        }
      });
      this.document = parser.parseFromString(cleanedXml, 'text/xml');
      this.root = this.document.documentElement;
      
      return this.root !== null && this.root.nodeName !== 'parsererror';
    } catch (e) {
      return false;
    }
  }

  // =========================
  // HELPER CORE (CHUẨN HOÁ)
  // =========================

  protected S(key: string): string | null {
    if (!this.root) return null;
    const node = this.findDescendant(this.root, key);
    if (!node) return null;
    return node.textContent ? node.textContent.trim() : '';
  }

  protected I(key: string): number | null {
    const s = this.S(key);
    if (s === null || s === '') return null;
    const v = parseInt(s, 10);
    return isNaN(v) ? null : v;
  }

  protected D(key: string): number | null {
    const s = this.S(key);
    if (s === null || s === '') return null;
    const v = parseFloat(s);
    return isNaN(v) ? null : v;
  }

  protected DT(key: string): Date | null {
    const s = this.S(key);
    return parseExactDate(s);
  }

  // =========================
  // HELPER ROOT CHECK
  // =========================

  protected isRoot(expectedLocalName: string): boolean {
    if (!this.root) return false;
    const rootName = (this.root.localName || this.root.tagName.split(':').pop())
      .replace(/^[\uFEFF]+/, '')
      .trim();
    return rootName === expectedLocalName;
  }

  // =========================
  // DOM TRAVERSAL HELPERS
  // =========================

  protected findDescendant(node: any, key: string): any {
    if (node.nodeType === 1) {
      const localName = node.localName || node.tagName.split(':').pop();
      if (localName === key) {
        return node;
      }
    }
    let child = node.firstChild;
    while (child) {
      const found = this.findDescendant(child, key);
      if (found) return found;
      child = child.nextSibling;
    }
    return null;
  }

  protected findDescendants(node: any, key: string): any[] {
    const results: any[] = [];
    this.collectDescendants(node, key, results);
    return results;
  }

  private collectDescendants(node: any, key: string, results: any[]) {
    if (node.nodeType === 1) {
      const localName = node.localName || node.tagName.split(':').pop();
      if (localName === key) {
        results.push(node);
      }
    }
    let child = node.firstChild;
    while (child) {
      this.collectDescendants(child, key, results);
      child = child.nextSibling;
    }
  }

  protected findChildren(node: any, key: string): any[] {
    const results: any[] = [];
    let child = node.firstChild;
    while (child) {
      if (child.nodeType === 1) {
        const localName = child.localName || child.tagName.split(':').pop();
        if (localName === key) {
          results.push(child);
        }
      }
      child = child.nextSibling;
    }
    return results;
  }

  // Helper inside specific elements
  protected getElementText(el: any, key: string): string | null {
    const node = this.findDescendant(el, key);
    if (!node) return null;
    return node.textContent ? node.textContent.trim() : '';
  }

  protected getElementInt(el: any, key: string): number | null {
    const s = this.getElementText(el, key);
    if (s === null || s === '') return null;
    const v = parseInt(s, 10);
    return isNaN(v) ? null : v;
  }

  protected getElementDecimal(el: any, key: string): number | null {
    const s = this.getElementText(el, key);
    if (s === null || s === '') return null;
    const v = parseFloat(s);
    return isNaN(v) ? null : v;
  }

  protected getElementDate(el: any, key: string): Date | null {
    const s = this.getElementText(el, key);
    return parseExactDate(s);
  }
}
